File: AtCoder/AtCoder - KYOCERA Programming Contest 2021（AtCoder Beginner Contest 200）/E - Patisserie ABC 2/solution.cpp
/********************************************************************
Problem Solving / Competitive Programming Template

Code by MatWhyTle(ingyu1008)
********************************************************************/

//#pragma GCC optimize("O3")
//#pragma GCC target("avx2")
//#pragma GCC optimize("unroll-loops")
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>

typedef unsigned long long ll;
typedef std::pair<int, int> pii;
typedef std::pair<ll, ll> pll;
typedef std::vector<pii> vpii;

ll mod = 1e9 + 7;

ll powmod(ll a, ll n)
{
    if (n == 0)
        return 1;
    if (n == 1)
        return a;
    ll b = powmod(a, n / 2);
    b = (b * b) % mod;
    if (n & 1)
        b = (b * a) % mod;
    return b;
}

ll gcd(ll a, ll b)
{
    if (a == 0)
        return b;
    if (b == 0)
        return a;
    if (a < b)
        a ^= b ^= a ^= b;
    while (b)
        b ^= a ^= b ^= a %= b;
    return a;
}

/********************************************************************
						End Of Template
********************************************************************/

int main(void)
{
    // std::cin.tie(0);
    // std::ios_base::sync_with_stdio(false);

    ll N, K;
    std::cin >> N >> K;

    std::vector<ll> v(3 * N + 1, 0);

    ll sum;

    for (ll i = 3; i <= 3 * N; i++)
    {
        ll now = ((((i - 1) * (i - 2)) / 2ULL) * (i - 3)) / 3ULL;
        v[i] = v[i - 1] + now;
        if (v[i] >= K)
        {
            sum = i - 1;
            break;
        }
    }

    std::cout << sum << "\n";
    //found sum
    ll beauty;

    std::vector<ll> vv(3 * N + 1, 0);
    for (ll i = std::max(1ULL, sum - 2 * N); i <= 3 * N; i++)
    {
        if (i + 2 > sum)
        {
            vv[i] = vv[i - 1];
        }
        else
        {
            ll now = sum - i - 1;
            vv[i] = vv[i - 1] + now;
            if (vv[i] >= K)
            {
                beauty = i - 1;
                break;
            }
        }
    }

    //beauty found

    std::cout << beauty << "\n";

    for (ll taste = 1; taste < sum - beauty; taste++)
    {
        if (v[sum - 1] + vv[beauty] == K)
        {
            std::cout << beauty << " " << taste << " " << sum - beauty - taste << "\n";
            break;
        }
        K--;
    }

    return 0;
}

File: AtCoder/AtCoder - KYOCERA Programming Contest 2021（AtCoder Beginner Contest 200）/E - Patisserie ABC 2/solution.py
N, K = map(int, input().split())

v = [0 for i in range(3*N + 1)]

for i in range(3, 3*N+1):
    now = (i-1)*(i-2)*(i-3)//6
    v[i] = v[i-1] + now

lo = 3
hi = 3*N

sm = 0
while lo <= hi:
    m = (lo + hi) // 2
    # print(f'{m} {v[m]} {K}')
    if v[m] < K:
        sm = max(sm, m)
        lo = m + 1
    else:
        hi = m - 1

sm += 1

vv = [0 for i in range(3*N + 1)]

for i in range(max(1, sm-2*N), 3*N+1):
    if i+2 > sm:
        vv[i] = vv[i-1]
    else :
        now = sm - i - 1
        vv[i] = vv[i-1] + now

lo = 1
hi = sm - 2
beauty = 0


while lo <= hi:
    m = (lo + hi) // 2
    if v[sm-1] + vv[m] < K:
        beauty = max(beauty, m)
        lo = m + 1
    else :
        hi = m - 1

for taste in range(1, sm - beauty) :
    if v[sm - 1] + vv[beauty] == K:
        print(f'{beauty} {taste} {sm - beauty - taste}')
        break
    K -= 1
